From: Yusuke Endoh <mame@ruby-lang.org>
Date: Fri, 13 Jun 2025 12:44:08 +0900
Subject: Reject `true && not true`

A command-call-like `not true` must be rejected after `&&` and `||`.

https://bugs.ruby-lang.org/issues/21337

----

Early backport of https://github.com/ruby/prism/pull/3585 due to its
bug severity, but missed 3.4.5 release.
---
 lib/prism/compiler.rb                  |  2 +-
 lib/prism/dispatcher.rb                |  2 +-
 lib/prism/dsl.rb                       |  2 +-
 lib/prism/inspect_visitor.rb           |  2 +-
 lib/prism/mutation_compiler.rb         |  2 +-
 lib/prism/node.rb                      |  2 +-
 lib/prism/reflection.rb                |  2 +-
 lib/prism/serialize.rb                 |  3 ++-
 lib/prism/visitor.rb                   |  2 +-
 prism/api_node.c                       |  3 +--
 prism/ast.h                            |  3 +--
 prism/config.yml                       |  1 +
 prism/diagnostic.c                     |  4 +++-
 prism/diagnostic.h                     |  1 +
 prism/node.c                           |  3 +--
 prism/prettyprint.c                    |  3 +--
 prism/prism.c                          | 12 ++++++++++++
 prism/serialize.c                      |  3 +--
 prism/templates/src/diagnostic.c.erb   |  1 +
 prism/token_type.c                     |  3 +--
 test/prism/errors/command_calls_31.txt | 17 +++++++++++++++++
 21 files changed, 51 insertions(+), 22 deletions(-)
 create mode 100644 test/prism/errors/command_calls_31.txt

diff --git a/lib/prism/compiler.rb b/lib/prism/compiler.rb
index b6771b3..a2e4b27 100644
--- a/lib/prism/compiler.rb
+++ b/lib/prism/compiler.rb
@@ -3,7 +3,7 @@
 =begin
 This file is generated by the templates/template.rb script and should not be
 modified manually. See templates/lib/prism/compiler.rb.erb
-if you are looking to modify the template
+if you are looking to modify the template.
 =end
 
 module Prism
diff --git a/lib/prism/dispatcher.rb b/lib/prism/dispatcher.rb
index 21fb14f..2482349 100644
--- a/lib/prism/dispatcher.rb
+++ b/lib/prism/dispatcher.rb
@@ -3,7 +3,7 @@
 =begin
 This file is generated by the templates/template.rb script and should not be
 modified manually. See templates/lib/prism/dispatcher.rb.erb
-if you are looking to modify the template
+if you are looking to modify the template.
 =end
 
 module Prism
diff --git a/lib/prism/dsl.rb b/lib/prism/dsl.rb
index 5047bd5..814dc66 100644
--- a/lib/prism/dsl.rb
+++ b/lib/prism/dsl.rb
@@ -3,7 +3,7 @@
 =begin
 This file is generated by the templates/template.rb script and should not be
 modified manually. See templates/lib/prism/dsl.rb.erb
-if you are looking to modify the template
+if you are looking to modify the template.
 =end
 
 module Prism
diff --git a/lib/prism/inspect_visitor.rb b/lib/prism/inspect_visitor.rb
index a35acbc..3cf43b3 100644
--- a/lib/prism/inspect_visitor.rb
+++ b/lib/prism/inspect_visitor.rb
@@ -3,7 +3,7 @@
 =begin
 This file is generated by the templates/template.rb script and should not be
 modified manually. See templates/lib/prism/inspect_visitor.rb.erb
-if you are looking to modify the template
+if you are looking to modify the template.
 =end
 
 module Prism
diff --git a/lib/prism/mutation_compiler.rb b/lib/prism/mutation_compiler.rb
index 02fa777..57a21e6 100644
--- a/lib/prism/mutation_compiler.rb
+++ b/lib/prism/mutation_compiler.rb
@@ -3,7 +3,7 @@
 =begin
 This file is generated by the templates/template.rb script and should not be
 modified manually. See templates/lib/prism/mutation_compiler.rb.erb
-if you are looking to modify the template
+if you are looking to modify the template.
 =end
 
 module Prism
diff --git a/lib/prism/node.rb b/lib/prism/node.rb
index 5147685..217e438 100644
--- a/lib/prism/node.rb
+++ b/lib/prism/node.rb
@@ -3,7 +3,7 @@
 =begin
 This file is generated by the templates/template.rb script and should not be
 modified manually. See templates/lib/prism/node.rb.erb
-if you are looking to modify the template
+if you are looking to modify the template.
 =end
 
 module Prism
diff --git a/lib/prism/reflection.rb b/lib/prism/reflection.rb
index 4297f4c..d32fd5d 100644
--- a/lib/prism/reflection.rb
+++ b/lib/prism/reflection.rb
@@ -3,7 +3,7 @@
 =begin
 This file is generated by the templates/template.rb script and should not be
 modified manually. See templates/lib/prism/reflection.rb.erb
-if you are looking to modify the template
+if you are looking to modify the template.
 =end
 
 module Prism
diff --git a/lib/prism/serialize.rb b/lib/prism/serialize.rb
index 932dd76..1bfcffa 100644
--- a/lib/prism/serialize.rb
+++ b/lib/prism/serialize.rb
@@ -3,7 +3,7 @@
 =begin
 This file is generated by the templates/template.rb script and should not be
 modified manually. See templates/lib/prism/serialize.rb.erb
-if you are looking to modify the template
+if you are looking to modify the template.
 =end
 
 require "stringio"
@@ -238,6 +238,7 @@ module Prism
         :expect_for_delimiter,
         :expect_ident_req_parameter,
         :expect_in_delimiter,
+        :expect_lparen_after_not,
         :expect_lparen_req_parameter,
         :expect_message,
         :expect_rbracket,
diff --git a/lib/prism/visitor.rb b/lib/prism/visitor.rb
index dfc2e4a..9dc3532 100644
--- a/lib/prism/visitor.rb
+++ b/lib/prism/visitor.rb
@@ -3,7 +3,7 @@
 =begin
 This file is generated by the templates/template.rb script and should not be
 modified manually. See templates/lib/prism/visitor.rb.erb
-if you are looking to modify the template
+if you are looking to modify the template.
 =end
 
 module Prism
diff --git a/prism/api_node.c b/prism/api_node.c
index f52cca8..3cf5468 100644
--- a/prism/api_node.c
+++ b/prism/api_node.c
@@ -2,8 +2,7 @@
 /* This file is generated by the templates/template.rb script and should not  */
 /* be modified manually. See                                                  */
 /* templates/ext/prism/api_node.c.erb                                         */
-/* if you are looking to modify the                                           */
-/* template                                                                   */
+/* if you are looking to modify the template                                  */
 /*----------------------------------------------------------------------------*/
 
 #line 2 "prism/templates/ext/prism/api_node.c.erb"
diff --git a/prism/ast.h b/prism/ast.h
index 86dee4d..13e12f1 100644
--- a/prism/ast.h
+++ b/prism/ast.h
@@ -2,8 +2,7 @@
 /* This file is generated by the templates/template.rb script and should not  */
 /* be modified manually. See                                                  */
 /* templates/include/prism/ast.h.erb                                          */
-/* if you are looking to modify the                                           */
-/* template                                                                   */
+/* if you are looking to modify the template                                  */
 /*----------------------------------------------------------------------------*/
 
 /**
diff --git a/prism/config.yml b/prism/config.yml
index cfa3f62..81765ce 100644
--- a/prism/config.yml
+++ b/prism/config.yml
@@ -101,6 +101,7 @@ errors:
   - EXPECT_FOR_DELIMITER
   - EXPECT_IDENT_REQ_PARAMETER
   - EXPECT_IN_DELIMITER
+  - EXPECT_LPAREN_AFTER_NOT
   - EXPECT_LPAREN_REQ_PARAMETER
   - EXPECT_MESSAGE
   - EXPECT_RBRACKET
diff --git a/prism/diagnostic.c b/prism/diagnostic.c
index 9842f15..5719ab7 100644
--- a/prism/diagnostic.c
+++ b/prism/diagnostic.c
@@ -8,7 +8,7 @@
 
 #include "prism/diagnostic.h"
 
-#define PM_DIAGNOSTIC_ID_MAX 319
+#define PM_DIAGNOSTIC_ID_MAX 320
 
 /** This struct holds the data for each diagnostic. */
 typedef struct {
@@ -192,6 +192,7 @@ static const pm_diagnostic_data_t diagnostic_messages[PM_DIAGNOSTIC_ID_MAX] = {
     [PM_ERR_EXPECT_FOR_DELIMITER]               = { "unexpected %s; expected a 'do', newline, or ';' after the 'for' loop collection", PM_ERROR_LEVEL_SYNTAX },
     [PM_ERR_EXPECT_IDENT_REQ_PARAMETER]         = { "expected an identifier for the required parameter", PM_ERROR_LEVEL_SYNTAX },
     [PM_ERR_EXPECT_IN_DELIMITER]                = { "expected a delimiter after the patterns of an `in` clause", PM_ERROR_LEVEL_SYNTAX },
+    [PM_ERR_EXPECT_LPAREN_AFTER_NOT]            = { "expected a `(` after `not`", PM_ERROR_LEVEL_SYNTAX },
     [PM_ERR_EXPECT_LPAREN_REQ_PARAMETER]        = { "expected a `(` to start a required parameter", PM_ERROR_LEVEL_SYNTAX },
     [PM_ERR_EXPECT_MESSAGE]                     = { "unexpected %s; expecting a message to send to the receiver", PM_ERROR_LEVEL_SYNTAX },
     [PM_ERR_EXPECT_RBRACKET]                    = { "expected a matching `]`", PM_ERROR_LEVEL_SYNTAX },
@@ -519,6 +520,7 @@ pm_diagnostic_id_human(pm_diagnostic_id_t diag_id) {
         case PM_ERR_EXPECT_FOR_DELIMITER: return "expect_for_delimiter";
         case PM_ERR_EXPECT_IDENT_REQ_PARAMETER: return "expect_ident_req_parameter";
         case PM_ERR_EXPECT_IN_DELIMITER: return "expect_in_delimiter";
+        case PM_ERR_EXPECT_LPAREN_AFTER_NOT: return "expect_lparen_after_not";
         case PM_ERR_EXPECT_LPAREN_REQ_PARAMETER: return "expect_lparen_req_parameter";
         case PM_ERR_EXPECT_MESSAGE: return "expect_message";
         case PM_ERR_EXPECT_RBRACKET: return "expect_rbracket";
diff --git a/prism/diagnostic.h b/prism/diagnostic.h
index 2b1aa9b..df4c62c 100644
--- a/prism/diagnostic.h
+++ b/prism/diagnostic.h
@@ -130,6 +130,7 @@ typedef enum {
     PM_ERR_EXPECT_FOR_DELIMITER,
     PM_ERR_EXPECT_IDENT_REQ_PARAMETER,
     PM_ERR_EXPECT_IN_DELIMITER,
+    PM_ERR_EXPECT_LPAREN_AFTER_NOT,
     PM_ERR_EXPECT_LPAREN_REQ_PARAMETER,
     PM_ERR_EXPECT_MESSAGE,
     PM_ERR_EXPECT_RBRACKET,
diff --git a/prism/node.c b/prism/node.c
index 642b8fc..2de43f5 100644
--- a/prism/node.c
+++ b/prism/node.c
@@ -2,8 +2,7 @@
 /* This file is generated by the templates/template.rb script and should not  */
 /* be modified manually. See                                                  */
 /* templates/src/node.c.erb                                                   */
-/* if you are looking to modify the                                           */
-/* template                                                                   */
+/* if you are looking to modify the template                                  */
 /*----------------------------------------------------------------------------*/
 
 #line 2 "prism/templates/src/node.c.erb"
diff --git a/prism/prettyprint.c b/prism/prettyprint.c
index 1beb3e5..7f64c3a 100644
--- a/prism/prettyprint.c
+++ b/prism/prettyprint.c
@@ -2,8 +2,7 @@
 /* This file is generated by the templates/template.rb script and should not  */
 /* be modified manually. See                                                  */
 /* templates/src/prettyprint.c.erb                                            */
-/* if you are looking to modify the                                           */
-/* template                                                                   */
+/* if you are looking to modify the template                                  */
 /*----------------------------------------------------------------------------*/
 
 #include "prism/prettyprint.h"
diff --git a/prism/prism.c b/prism/prism.c
index 6fb40e3..98a1c12 100644
--- a/prism/prism.c
+++ b/prism/prism.c
@@ -19725,6 +19725,18 @@ parse_expression_prefix(pm_parser_t *parser, pm_binding_power_t binding_power, b
             pm_arguments_t arguments = { 0 };
             pm_node_t *receiver = NULL;
 
+            if (!accepts_command_call && !match1(parser, PM_TOKEN_PARENTHESIS_LEFT)) {
+                if (parser->current.type == PM_TOKEN_PARENTHESIS_LEFT_PARENTHESES) {
+                    pm_parser_err(parser, parser->previous.end, parser->previous.end + 1, PM_ERR_EXPECT_LPAREN_AFTER_NOT);
+                } else if (parser->current.type == PM_TOKEN_NEWLINE) {
+                    parser_lex(parser);
+                    pm_parser_err_current(parser, PM_ERR_EXPECT_LPAREN_AFTER_NOT);
+                } else {
+                    pm_parser_err_current(parser, PM_ERR_EXPECT_LPAREN_AFTER_NOT);
+                }
+                return (pm_node_t *) pm_missing_node_create(parser, parser->current.start, parser->current.end);
+            }
+
             accept1(parser, PM_TOKEN_NEWLINE);
 
             if (accept1(parser, PM_TOKEN_PARENTHESIS_LEFT)) {
diff --git a/prism/serialize.c b/prism/serialize.c
index 9aba958..1075358 100644
--- a/prism/serialize.c
+++ b/prism/serialize.c
@@ -2,8 +2,7 @@
 /* This file is generated by the templates/template.rb script and should not  */
 /* be modified manually. See                                                  */
 /* templates/src/serialize.c.erb                                              */
-/* if you are looking to modify the                                           */
-/* template                                                                   */
+/* if you are looking to modify the template                                  */
 /*----------------------------------------------------------------------------*/
 
 #include "prism.h"
diff --git a/prism/templates/src/diagnostic.c.erb b/prism/templates/src/diagnostic.c.erb
index c0ec696..d171f8e 100644
--- a/prism/templates/src/diagnostic.c.erb
+++ b/prism/templates/src/diagnostic.c.erb
@@ -184,6 +184,7 @@ static const pm_diagnostic_data_t diagnostic_messages[PM_DIAGNOSTIC_ID_MAX] = {
     [PM_ERR_EXPECT_FOR_DELIMITER]               = { "unexpected %s; expected a 'do', newline, or ';' after the 'for' loop collection", PM_ERROR_LEVEL_SYNTAX },
     [PM_ERR_EXPECT_IDENT_REQ_PARAMETER]         = { "expected an identifier for the required parameter", PM_ERROR_LEVEL_SYNTAX },
     [PM_ERR_EXPECT_IN_DELIMITER]                = { "expected a delimiter after the patterns of an `in` clause", PM_ERROR_LEVEL_SYNTAX },
+    [PM_ERR_EXPECT_LPAREN_AFTER_NOT]            = { "expected a `(` after `not`", PM_ERROR_LEVEL_SYNTAX },
     [PM_ERR_EXPECT_LPAREN_REQ_PARAMETER]        = { "expected a `(` to start a required parameter", PM_ERROR_LEVEL_SYNTAX },
     [PM_ERR_EXPECT_MESSAGE]                     = { "unexpected %s; expecting a message to send to the receiver", PM_ERROR_LEVEL_SYNTAX },
     [PM_ERR_EXPECT_RBRACKET]                    = { "expected a matching `]`", PM_ERROR_LEVEL_SYNTAX },
diff --git a/prism/token_type.c b/prism/token_type.c
index ba2d751..e1e12f0 100644
--- a/prism/token_type.c
+++ b/prism/token_type.c
@@ -2,8 +2,7 @@
 /* This file is generated by the templates/template.rb script and should not  */
 /* be modified manually. See                                                  */
 /* templates/src/token_type.c.erb                                             */
-/* if you are looking to modify the                                           */
-/* template                                                                   */
+/* if you are looking to modify the template                                  */
 /*----------------------------------------------------------------------------*/
 
 #include <string.h>
diff --git a/test/prism/errors/command_calls_31.txt b/test/prism/errors/command_calls_31.txt
new file mode 100644
index 0000000..c34386c
--- /dev/null
+++ b/test/prism/errors/command_calls_31.txt
@@ -0,0 +1,17 @@
+true && not true
+            ^~~~ expected a `(` after `not`
+            ^~~~ unexpected 'true', expecting end-of-input
+
+true || not true
+            ^~~~ expected a `(` after `not`
+            ^~~~ unexpected 'true', expecting end-of-input
+
+true && not (true)
+           ^ expected a `(` after `not`
+            ^ unexpected '(', expecting end-of-input
+
+true && not
+true
+^~~~ expected a `(` after `not`
+^~~~ unexpected 'true', expecting end-of-input
+
